### ISSUE TITLE
Webmin RPM installation assumes SysV init.d even on systemd distributions

### ISSUE THREAD

**Issue by ElScotto** [webmin/webmin#1556]
The Webmin RPM installation always drops in an "init.d" script regardless of whether the system uses "systemd" or not.
Whilst systemd currently supports legacy init.d scripts, this may not always be the case.
So, the install scriptlet in the RPM should detect whether the system is using systemd, upstart, or traditional init.d startup scripts and process accordingly.

I've attached a copy of my webmin.service systemd unit file ("`/etc/systemd/system/webmin.service`").
I've also attached the override file that I use for configuring the Webmin PID file location ("`/etc/systemd/system/webmin.service.d/override.conf`")
[override.conf](https://github.com/webmin/webmin/files/7645122/override.conf.txt)
[webmin.service](https://github.com/webmin/webmin/files/7645123/webmin.service.txt)

The RPM installation script can drop in the basic "webmin.service" file, and treat the "override.conf" file as being part of the package but "volatile", so that an RPM verify won't trip over it if the user needs to customise it.

**Comment by jcameron** [comment]
We do this intentionally so that a single RPM can be compatible with the maximum number of different Linux distributions. So far, I haven't seen an RPM-based distro that doesn't support init scripts..

**Comment by ElScotto** [comment]
Fair enough.  I don't have an issue with a single, monolithic RPM that tries to cater for maximum compatibility.
The frustration is that the RPM scriptlets are not validating the environment, and just blindly "doing stuff".
systemd will handle legacy "`init.d/webmin`" scripts (for now), and will quite happily invoke them if there's no matching "`webmin.service`" file, but does not like it when restarts occur outside its knowledge.
In this instance, using the "`init.d/webmin { start | stop | restart }`" command rather than "`systemctl { start | stop | restart } webmin`" causes systemd to mark the service as "failed" because the contents of the PIDfile ("`/var/webmin/miniserv.pid`") changed and systemd still thinks the old PID (which is no longer running) is the real PID.

Now, could systemd handle this better?  You betcha 😒  - and I'll see what the systemd folk can offer in this regard 🙄 .
However, just blindly "doing stuff" without consideration of the execution environment is ... sub-optimal. :wink:

**Comment by jcameron** [comment]
Will the `systemctl start` command work for init scripts though? Alternately, would a command like `service webmin restart` work?

**Comment by ElScotto** [comment]
> Will the `systemctl start` command work for init scripts though? Alternately, would a command like `service webmin restart` work?

Yes, it will work with init scripts but **only** as a last resort; if there's a _unit.service_ file (eg `webmin.service`) that matches the name of a script in init.d (eg `webmin`), systemd will always use the service file in preference.

Here's what systemd reports when using the init script (via `systemctl status webmin`):
[webmin-systemctl-status-sysv-init.txt](https://github.com/webmin/webmin/files/7659968/webmin-systemctl-status-sysv-init.txt)

And here's how it looks when the `webmin.service` file is in place:
[webmin-systemctl-status-systemd-service.txt](https://github.com/webmin/webmin/files/7659967/webmin-systemctl-status-systemd-service.txt)

In this instance, I temporarily moved my `webmin.service` file out of the way so that systemd would only find the SysV init script.
What systemd does is actually build a "dynamic service" via "`systemd-sysv-generator`" (cf: [man:systemd-sysv-generator(8)](https://man7.org/linux/man-pages/man8/systemd-sysv-generator.8.html)), which is actually pretty bloody clever.
However, for finer-grained control and integration with systemd, a unit.service file is far superior.

So whilst the init.d script can be used in systemd, it won't be used if there's already an equivalent `.service` file installed.
Thus when the RPM restarts webmin (eg using `init.d/webmin restart`) after the upgrade, systemd doesn't see the change to the PID etc because it happened "outside" of the service file.

**Comment by jcameron** [comment]
So as a small improvement, we could have Webmin's install script run `service webmin restart` rather than `/etc/init.d/webmin restart` , since the `service` command is pretty commonly available.

**Comment by ElScotto** [comment]
> So as a small improvement, we could have Webmin's install script run `service webmin restart` rather than `/etc/init.d/webmin restart` , since the `service` command is pretty commonly available.

Sounds logical and quite reasonable, however after trying it, I can report that whilst the `service webmin status` and `service webmin restart` commands worked in a systemd-enabled system, unfortunately (for whatever reason) systemd **didn't** see the PID change (and didn't re-read the PID file), thus thinks the service has failed, even though it is actually running.

[webmin-service-status-restart-under-systemd.txt](https://github.com/webmin/webmin/files/7667924/webmin-service-status-restart-under-systemd.txt)

So, in the attached file, you can see that the original PID reported by both "status" commands is the same (ie pid 30450)
However after issuing "`service webmin restart`", a new PID is set (ie pid 28836) but systemd doesn't see that and assumes that the old PID is still the main PID.

Question: The "`/etc/webmin/reload`" command effectively issues a "`kill -USR1`" to the main PID (ie the Perl "loader"); does that actually force Perl to reload the "`miniserv.pl`" code and restart or does it just trigger the Perl script to re-read its configuration file and continues to use the in-memory (old version) of the "`miniserv.pl`" code?
If it's the former, then using the `reload` option would be better than `restart` because the PID doesn't change.

(The fact that systemd is not re-reading the PID file after processing the `restart` sub-command is annoying and possibly a bug; it seems to be conflating 'restart' with 'reload' which I feel is wrong; 'restart' to my mind is a full stop followed by a full start which will invariably result in a new PID whereas I interpret 'reload' as just that; a reload of its configuration/options from disc)

**Comment by jcameron** [comment]
I think we can fix this by not making the init script part of the RPM, but instead have it generated by the install process using the appropriate boot system. However, the tricky part will be transitioning existing installs without leaving them in a broken state.. 

I will look into this, and update this ticket with details. Apparently this is a requirement on CentOS 9 as it lacks init.d support entirely.

**Comment by iliarostovtsev** [comment]
For _systemd_, the real problem seems to be checking the **process origin**. If Webmin is started using `/etc/webmin/start` or `service webmin start` than any further `systemctl [status|restart|stop] webmin` simply won't work on CentOS Stream 9.

However, if Webmin process is started using `systemctl start webmin` command, any further commands like `systemctl [restart|stop|status] webmin` **work as expected** (PID changes and everything). 

> I will look into this, and update this ticket with details. 

Perhaps, this can be worked around? I am afraid that this change could potentially break existing installs and systems without _systemd_ and/or bring up new issues on the dawn of Virtualmin 7 release! 

> Apparently this is a requirement on CentOS 9 as it lacks init.d support entirely.

Not at all. My suggestions which I initially reported to our group chat were not complete in relation to installing Webmin on CentOS Stream 9. After taking a deeper look, I found that current version of Webmin can be easily installed on CentOS Stream 9 as easy as:

```
dnf install chkconfig perl-lib
systemctl daemon-reload
dnf install https://download.webmin.com/devel/rpm/webmin-current.rpm
```

It appeared that the source of the problem was missing `chkconfig` package. If you check **`/etc/rc.d/init.d/README`** file on CentOS Stream 9 it clearly says that **init.d support is _not gone_**:

```
You are looking for the traditional init scripts in /etc/rc.d/init.d,
and they are gone?

Here's an explanation on what's going on:

You are running a systemd-based OS where traditional init scripts have
been replaced by native systemd services files. Service files provide
very similar functionality to init scripts. To make use of service
files simply invoke "systemctl", which will output a list of all
currently running services (and other units). Use "systemctl
list-unit-files" to get a listing of all known unit files, including
stopped, disabled and masked ones. Use "systemctl start
foobar.service" and "systemctl stop foobar.service" to start or stop a
service, respectively. For further details, please refer to
systemctl(1).

Note that traditional init scripts continue to function on a systemd
system. An init script /etc/rc.d/init.d/foobar is implicitly mapped
into a service unit foobar.service during system initialization.
```

----

Don't get me wrong, I don't mind against creating native _systemd_ service file. I am just sharing this info so you'd know.

What is interesting, that we perhaps could solve this for now by a better way of manipulating automatically generated `/run/systemd/generator.late/webmin.service` file by `systemd-sysv-generator` based on LSB headers in _initd_ script?

Although for whatever I tried on CentOS Stream 9 I could not overcome the problem, where `systemctl [status|restart|stop|start] webmin` would work if Webmin was started with either `/etc/webmin/start` or `/etc/init.d/webmin start` and it seems that **`systemd`** simply doesn't know anything about it if started elsewhere.

So the solution to the problem could be is to make sure that we always call `systemctl [status|restart|stop|start] webmin` on _systemd_ systems, as it works perfectly fine with our old _initd_ script. Presumably, our initial _initd_ _webmin_ script should be edited to have a different paths for _START_, _STOP_ and _RELOAD_ commands, where initial commands (`/etc/webmin/[restart|start|stop`]) would be symlinks to either original command or will call `systemctl [status|restart|stop|start] webmin` directly in case of _systemd_ systems. Besides we need to add support for `force-reload` as well, which will call `stop --kill`.

**Comment by jcameron** [comment]
I'm going to have to do some more research into the best way to implement this..

**Comment by tneo** [comment]
It probably is me. On an openSUSE 15.3 server with Webmin installed any service I want start/stop wants to execute an init.d script, but none are existent. I stopped Webmin and restarted it using systemctl as I understood from this ticket that should trigger webmin to execute systemctl and not init.d scripts. Do I need to manually add init.d scripts/ symlinks?

**Comment by iliajie** [comment]
> Do I need to manually add init.d scripts/ symlinks?

No, I will look into adding a better support for the latest openSUSE systems.

**Comment by tneo** [comment]
I noticed with several services I was able to change the start/ stop command for the service to use the systemctl command. This works, however Webmin does not recognize that Apache is running and will always show "start Apache" in the top-right corner.

**Comment by tneo** [comment]
I did change the commands for Apache (and MariaDB for that matter), but Webmin does not recognize the Apache status.

**Comment by iliajie** [comment]
What about PID file and config file/directory?

**Comment by tneo** [comment]
Aha. Updating the PID to the correct file (on openSUSE 15.3 that is /var/run/httpd.pid) makes Webmin recognize the Apache status.

**Comment by iliajie** [comment]
What is the output of `cat /etc/os-release` on your system?

**Comment by tneo** [comment]
`cat /etc/os-release
NAME="openSUSE Leap"
VERSION="15.3"
ID="opensuse-leap"
ID_LIKE="suse opensuse"
VERSION_ID="15.3"
PRETTY_NAME="openSUSE Leap 15.3"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:leap:15.3"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/`

**Comment by iliajie** [comment]
Alright, upcoming Webmin 2.010 will have much greater support for openSUSE Leap. The devel package which can be installed right now can be [downloaded here](https://ilia.engineer/webmin-2.001-67.noarch.rpm).

It will only work if you completely reinstall Webmin or delete all config files under Webnin modules, i.e. `/etc/webmin/*/config` files.

Please feel free to give it a try and let me know how it works for you. Backup `/etc/webmin` before trying!

**Comment by tneo** [comment]
I wanted to give the build a try, but the link you provided throws a 404.

**Comment by tneo** [comment]
wget https://ilia.engineer/webmin-2.001-68.noarch.rpm
> --2022-11-28 09:06:55--  https://ilia.engineer/webmin-2.001-68.noarch.rpm
> Resolving ilia.engineer (ilia.engineer)... 188.114.96.3, 188.114.97.3, 2a06:98c1:3120::, ...
> Connecting to ilia.engineer (ilia.engineer)|188.114.96.3|:443... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2022-11-28 09:06:55 ERROR 404: Not Found.

**Comment by iliajie** [comment]
Please check GitHub Releases page. Webmin 2.010 is already out.
